include/drivers/video.h

```diff
@@ -0,0 +1,556 @@
+		(const struct video_driver_api *)dev->driver_api;
+
+	__ASSERT(api->get_caps, "get_caps must be implemented by driver");
+
+	return api->get_caps(dev, ep, caps);
+}
+
+/**
+ * @brief Set the value of a control.
+ *
+ * This set the value of a video control, value type depends on control ID, and
+ * must be interpreted accordingly.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cid Control ID.
+ * @param value Pointer to the control value.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -ENOTSUP If format is not supported.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_set_ctrl(struct device *dev, unsigned int cid,
+				 void *value)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->set_ctrl == NULL) {
+		return -ENOTSUP;
+	}
+
+	return api->set_ctrl(dev, cid, value);
+}
+
+/**
+ * @brief Get the current value of a control.
+ *
+ * This retrieve the value of a video control, value type depends on control ID,
+ * and must be interpreted accordingly.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cid Control ID.
+ * @param value Pointer to the control value.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -ENOTSUP If format is not supported.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_get_ctrl(struct device *dev, unsigned int cid,
+				 void *value)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->get_ctrl == NULL) {
+		return -ENOTSUP;
+	}
+
+	return api->get_ctrl(dev, cid, value);
+}
+
+/**
+ * @brief Register/Unregister event callback for a video endpoint.
+ *
+ * Register a endpoint event handler which is called on frame completion,
+ * error, etc (cf video_event enum). There is no guarentee on the
+ * context, callback can be executed in IRQ or Thread context. Registering a
+ * NULL callback unregisters any previously registered callback.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param ep Endpoint ID.
+ * @param cb Event callback function.
+ * @param user_data Pointer to user data, passed back event callback
+ *
+ * @retval 0 Is successful, -ERRNO code otherwise.
+ */
+static inline int video_register_event_callback(struct device *dev,
+						enum video_endpoint_id ep,
+						video_event_callback_t cb,
+						void *user_data)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->register_event_callback == NULL) {
+		return -ENOTSUP;
+	}
+
+	return api->register_event_callback(dev, ep, cb, user_data);
+}
+
+/**
+ * @brief Allocate video buffer.
+ *
+ * @param size Size of the video buffer.
+ *
+ * @retval pointer to allocated video buffer
+ */
+struct video_buffer *video_buffer_alloc(size_t size);
```

Comment:
If allocation fails does this return a NULL? Does this block the calling context until the allocation can happen?

Reply:
It does not block and return NULL if memory is not available.